File: ADT/dfs_bfs/2101.py
# Google

# DFS
from collections import defaultdict


class Solution:
    def DFS(self, u, visited, adj):
        visited.add(u)
        for v in adj[u]:
            if v not in visited:
                self.DFS(v, visited, adj)

    def maximumDetonation(self, bombs: List[List[int]]) -> int:
        n = len(bombs)
        adj = defaultdict(list)
        for i in range(n):
            for j in range(n):
                if i == j:
                    continue
                x1, y1, r1 = bombs[i]
                x2, y2, r2 = bombs[j]

                distance = (x2 - x1) ** 2 + (y2 - y1) ** 2
                if r1 * r1 >= distance:
                    adj[i].append(j)

        result = 0
        visited = set()
        for i in range(n):
            self.DFS(i, visited, adj)
            count = len(visited)
            result = max(result, count)
            visited.clear()
        return result


# BFS
from collections import defaultdict, deque


class Solution:
    def BFS(self, u, adj):
        visited = set()
        que = deque()
        que.append(u)
        visited.add(u)
        while que:
            temp = que.popleft()
            for v in adj.get(temp, []):
                if v not in visited:
                    que.append(v)
                    visited.add(v)
        return len(visited)

    def maximumDetonation(self, bombs: List[List[int]]) -> int:
        n = len(bombs)
        adj = defaultdict(list)
        for i in range(n):
            for j in range(n):
                if i == j:
                    continue
                x1, y1, r1 = bombs[i]
                x2, y2, r2 = bombs[j]

                distance = (x2 - x1) ** 2 + (y2 - y1) ** 2
                if r1 * r1 >= distance:
                    adj[i].append(j)

        result = 0
        for i in range(n):
            count = self.BFS(i, adj)
            result = max(result, count)
        return result

File: ADT/dfs_bfs/2101.js

//DFS

var maximumDetonation = function (bombs) {
    const DFS = (u, visited, adj) => {
        visited.add(u)
        for (const v of (adj.get(u) || [])) {
            if (!visited.has(v)) {
                DFS(v, visited, adj)
            }
        }
    }
    const n = bombs.length
    const adj = new Map()
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            if (i === j) continue
            const x1 = bombs[i][0]
            const y1 = bombs[i][1]
            const r1 = bombs[i][2]

            const x2 = bombs[j][0]
            const y2 = bombs[j][1]
            const r2 = bombs[j][2]

            const distance = (x2 - x1) ** 2 + (y2 - y1) ** 2

            if (r1 * r1 >= distance) {
                if (!adj.has(i)) adj.set(i, [])
                adj.get(i).push(j)
            }
        }
    }
    let result = 0
    const visited = new Set()
    for (let i = 0; i < n; i++) {
        visited.clear()
        DFS(i, visited, adj)
        result = Math.max(result, visited.size)
    }
    return result
};


// BFS

var maximumDetonation = function (bombs) {
    const BFS = (u, adj) => {
        const visited = new Set()
        const queue = []
        queue.push(u)
        visited.add(u)
        while (queue.length > 0) {
            const temp = queue.shift()
            for (const v of (adj.get(temp) || [])) {
                if (!visited.has(v)) {
                    queue.push(v)
                    visited.add(v)
                }
            }
        }
        return visited.size
    }
    const n = bombs.length
    const adj = new Map()
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            if (i === j) continue
            const x1 = bombs[i][0]
            const y1 = bombs[i][1]
            const r1 = bombs[i][2]

            const x2 = bombs[j][0]
            const y2 = bombs[j][1]
            const r2 = bombs[j][2]

            const distance = (x2 - x1) ** 2 + (y2 - y1) ** 2

            if (r1 * r1 >= distance) {
                if (!adj.has(i)) adj.set(i, [])
                adj.get(i).push(j)
            }
        }
    }
    let result = 0
    for (let i = 0; i < n; i++) {
        const count = BFS(i, adj)
        if (count > result) result = count
    }
    return result
};